shakenfist/external_api/instance.py

```diff
@@ -91,7 +91,8 @@ def get(self, all=False):
     @api_base.requires_namespace_exist
     def post(self, name=None, cpus=None, memory=None, network=None, disk=None,
              ssh_key=None, user_data=None, placed_on=None, namespace=None,
-             video=None, uefi=False, configdrive=None):
+             video=None, uefi=False, configdrive=None, affinity=None,
+             tags=None):
```

Comment:
What is the difference between a tag and instance metadata? Should this instead be a more generic way of setting metadata on instance creation, that then appears via the metadata APIs?

Reply:
Tags are more controlled in there format ie. a list of strings. Unless we want to uses keys in metadata as tags? That seemed a bit too overloaded but I could be convinced.

Reply:
I was thinking we could have a metadata key for the instance called "tags", which requires that there be a list of values as its value? I guess I was also thinking that I should be able to set _any_ metadata key on the instance create call, not just tags?

Reply:
I think there is two suggestions here? Move `tags` to metadata, and secondly add metadata to the `instance create` API?

The reasoning behind making tags a separate parameter, is that tags should be static after instance start. If the tags are changed whilst the instance is running, should the instance move nodes?

Reply:
Oh I agree that tags / affinity are static values. I hadn't thought about the implications of that if we smash them into the existing metadata. It just feels weird to me to have two parallel metadata systems. I guess there are other forms of metadata which aren't updated -- for example if I implemented a default ssh key in namespace metadata, then you'd get the value at the time of instance create and it wouldn't be updated.

I guess that's the other thing here, if it was metadata then we could do metadata inheritance -- set these on the namespace if you want them to apply to all your instances for example.

Reply:
I think SSH keys are a prime target for an inheritance system. So there will be a number of parameters that could be inherited in such a system: `ssh_key`, `metadata`, `tags`. Tags don't need to be in metadata for that.

I think the most powerful argument to move `tags` to `metadata` is: what if we hate affinity using tags for scheduling and want to remove it? It would be better to not have `tags` in the API.

[moves to sitting on fence]